proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,200 @@
+
+**Document Metadata**
+
+**Author:** solomonkinard
+
+**Sponsoring Browser:** Chromium, Safari, Firefox
+
+**Contributors:** oliverdunk, xeenon, carlosjeurissen, hanguokai
+
+**Created:** 2024-04-05
+
+**Related Issues:**
+* https://crbug.com/893175
+* https://github.com/w3c/webextensions/issues/229
+
+## Motivation
+
+### Objective
+
+Extension developers will be able to supply and define a set of icons to be used
+in the event that user has expressed the preference for a page that has a dark
+theme.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari rely on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome relies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+manifest.json
+```
+"icon_variants": [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png",
+  },
+  {
+    "16": "dark16.png",
+    "32": "dark32.png",
+    "color_scheme": "dark",
+  },
+  {
+    "16": "light16.png",
+    "32": "light32.png",
+    "color_scheme": ["dark", "light"]
+  }
+]
+```
+
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const createProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties};
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
```

Comment:
Considering some browsers, while in light mode, might want to use the dark icon variant for better readability if a darker theme is used. I would suggest to rephrase it to something like this:

```suggestion
The `dark` icon variant can be used by browsers to improve readability of the icon. Often this will be when the os/browser color scheme is dark (often referred to as "dark mode") but also when the browser uses a darker shade theme which has a background on which the dark icon variant would be more readable.
```

Reply:
I agree with the general direction, but could we say something like "situations where the browser may choose to use this include", and then have both as bullet points? I'd like to avoid anything that gives the impression it is dictating when the browser picks them.

Reply:
@oliverdunk Sure! In general that would make it more readable as well. In any case the intention of the suggestion was to give potential examples. The whole purpose of using `color_scheme` and the syntax we are after is to go for the best contract with the intentions of the developer with the decisions made by the browser.